### PR TITLE
Add Buildcheck events support + BuildSubmissionStarted

### LINKED ISSUE #796: Add suport for the new binlog version (BuildSubmissionStartedEventArgs added)
### Context

https://github.com/dotnet/msbuild/pull/10424 is incrementing the log version. Let's add support for that version.

### Notes

It'd as well be nice to handle https://github.com/dotnet/msbuild/issues/10146 at the same time - by just capturing the steps thatt needed to be taken + linkid the respective 2 PRs (MSBuild and MSBuildStructuredLog)

### PR DESCRIPTION
**Pull request by YuliiaKovalova**
## Context
Add BuildCheck events (https://github.com/dotnet/msbuild/commit/ac76b77ef77a47d0926d214c3310ce46b10acb4a#diff-d88dc5683a31c5508ad8efa3aeecdfb56949fe6c0d234ff233afee954f93deb6) support and BuildSubmissionStarted (https://github.com/dotnet/msbuild/pull/10424)

### CODE REVIEW

**Comment by KirillOsenkov** [comment]
Ping me when this is ready for review!

**Comment by YuliiaKovalova** [comment]
@KirillOsenkov , it's ready for review, please take a look!

File: src/StructuredLogger/BinaryLogger/BuildEventArgsReader.cs [hunk 1]
```diff
@@ -1722,6 +1768,20 @@ private IEnumerable ReadProjectItems()
...
+
+            var list = new string[count];
+            for (int i = 0; i < count; i++)
+            {
+                list[i] = ReadDeduplicatedString();
+            }
+
+            return list;
+        }
+
```
Comment (KirillOsenkov) [inline src/StructuredLogger/BinaryLogger/BuildEventArgsReader.cs:1783]: In C# there should be never two consecutive empty lines of code

File: src/StructuredLogger/BinaryLogger/BuildSubmissionStartedEvent.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.Build.Framework;
+
+namespace StructuredLogger.BinaryLogger
+{
+    internal class BuildSubmissionStartedEvent() : BuildStatusEventArgs("", null, null, DateTime.UtcNow)
```
Comment (KirillOsenkov) [inline src/StructuredLogger/BinaryLogger/BuildSubmissionStartedEvent.cs:7]: it's good to specify argument names for literals such as strings, null, true and false, so when you read the code you know what parameter they map to

**Comment by KirillOsenkov** [comment]
We should also update our copy of the Writer here:
https://github.com/KirillOsenkov/MSBuildStructuredLog/blob/main/src/StructuredLogger/BinaryLogger/BuildEventArgsWriter.cs#L158

It's rarely used (and maybe we should delete it?) but that's a separate conversation I think

@JanKrivanek if you could review this PR too when you get a chance.

Also didn't we want to write down some instructions on updating the viewer copy of the reader/writer? I can't remember if we already have these somewhere or whether we still need to do it.

Appreciate the help @YuliiaKovalova!

File: src/StructuredLogger/BinaryLogger/BuildEventArgsReader.cs [hunk 3]
```diff
@@ -1245,6 +1248,49 @@ private BuildEventArgs ReadEnvironmentVariableReadEventArgs()
             return e;
         }
 
+        private BuildEventArgs ReadBuildCheckTracingEventArgs()
+        {
+            var fields = ReadBuildEventArgsFields(readImportance: true);
+            var rawTracingData = ReadStringDictionary() ?? new Dictionary<string, string>();
+
+            var e = new BuildCheckTracingEventArgs(rawTracingData.ToDictionary(
```
Comment (JanKrivanek) [inline src/StructuredLogger/BinaryLogger/BuildEventArgsReader.cs:1256]: Is this class already known to viewer?

**Comment by JanKrivanek** [comment]
Writer is needed - it's used for redacting functionality.

Documenting the viewer updating steps is as well still needed - I still didn't get a chance to fix it https://github.com/dotnet/msbuild/issues/10146

I'll surely and gladly review this - great thanks @YuliiaKovalova for adding all the missing types!!

**Review by JanKrivanek — APPROVED** [review]
Looks good!

**Comment by KirillOsenkov** [comment]
Just needs the writer changes.

**Comment by KirillOsenkov** [comment]
Also do we have a unit-test like BinaryLoggerRoundtrip that reads and writes the same file? I think we do have something... Would be nice to make sure we didn't "forget" the writer changes somehow.

File: src/StructuredLogger/BinaryLogger/BuildEventArgsReader.cs (hunk 3 above)
Comment (YuliiaKovalova) [reply src/StructuredLogger/BinaryLogger/BuildEventArgsReader.cs:1256]: I have added it as a part of this PR
https://github.com/KirillOsenkov/MSBuildStructuredLog/pull/797/files#diff-3280c719e5b13a87456db15d9462ea10354769ef6fc8d731c436ac74a183bfa7:~:text=internal%20sealed%20class-,BuildCheckTracingEventArgs,-(Dictionary%3C

Comment (YuliiaKovalova) [reply src/StructuredLogger/BinaryLogger/BuildEventArgsReader.cs:1256]: image

**Comment by YuliiaKovalova** [comment]
> BuildEventArgsWriter.cs#L158

We have this class:https://github.com/dotnet/msbuild/blob/5c934f81bcdf63168050f7083c4fd5b09faab9cc/src/Build.UnitTests/BuildEventArgsSerialization_Tests.cs but I see some missed recent events. 
I would suggest to add a note to https://github.com/dotnet/msbuild/blob/a03ee4ba9c5e71a361d86a6ef7923c44a0dc5f39/src/Build/Logging/BinaryLogger/BinaryLogger.cs#L81 to keep BuildEventArgsWriter(s) in msbuild and in Viewer consistent.

**Review by KirillOsenkov — APPROVED** [review]
👍🏻

**Comment by KirillOsenkov** [comment]
Thanks!

**Comment by KirillOsenkov** [comment]
Published:
 * https://github.com/KirillOsenkov/MSBuildStructuredLog/releases/tag/v2.2.317
 * https://www.nuget.org/packages/MSBuild.StructuredLogger/2.2.317
 * https://www.nuget.org/packages/binlogtool/1.0.17

**Pull request merged**